doc/architectural/central-data-structures.md

```diff
@@ -0,0 +1,145 @@
+\ingroup module_hidden
+
+\page central-data-structures Central Data Structures
+
+# Central Data Structures
+
+The following is some light technical documentation of the major data structures
+used in the input transformation to Intermediate Representation (IR) inside
+CBMC and the assorted CProver tools.
+
+## goto_modelt
+
+The `goto_modelt` is the top-level data structure that CBMC (and the other
+tools) use for holding the GOTO intermediate representation. The following
+diagram is a simplified view of how the data structures relate to each other -
+
+```mermaid
+erDiagram
+  goto_modelt {
+    symbol_tablet symbol_table
+    goto_functionst goto_functions
+  }
+  goto_modelt ||--|| symbol_tablet : ""
+  goto_modelt ||--|| goto_functionst : ""
+  symbol_tablet {
+    symbol_map symbols
+  }
+  symbol_tablet ||--o{ symbolt : ""
+  symbolt {
+    string name
+  }
+  goto_functionst {
+    function_map functions
+  }
+  goto_functionst ||--o{ goto_functiont : ""
+  goto_functiont {
+    goto_programt body
+    ordered_collection_of parameter_identifiers
+  }
+  goto_functiont ||--|| goto_programt : ""
+  goto_programt {
+    ordered_collection_of instructions
+  }
+  goto_programt ||--o{ instructiont : ""
+  instructiont {
+    enumeration instruction_type
+    goto_program_codet code
+    boolean_expression guard
+    source_locationt source_location
+  }
+  instructiont ||--|| goto_program_instruction_typet : ""
+  instructiont ||--o| goto_program_codet : ""
+  instructiont ||--o| source_locationt : ""
+  goto_program_codet ||--o| source_locationt : ""
+```
+
+A `goto_modelt` is effectively a pair, consisting of:
+
+* A list of GOTO functions,
+* A symbol table containing symbol references for the symbols contained in the GOTO functions.
+
+In pseudocode, the type looks this:
+
+```js
+type goto_modelt {
+  type goto_functionst = list<goto_functiont>
+  type symbol_tablet = map<identifier, symbolt>
+}
+```
+
+The abstract interface of `goto_modelt` is outlined in the file
+[`src/goto-programs/abstract_goto_model.h`](../../src/goto-programs/abstract_goto_model.h).
+
+## goto_functiont
+
+A `goto_functiont` is also defined as a pair. It's designed to represent a function
+at the goto level, and effectively it's the following data type (in pseudocode):
+
+```js
+type goto_functiont {
+    goto_programt body
+    list<identifiers> parameters
+}
+```
+
+The `goto_programt` denoting the `body` of the function will be the subject of
+a more elaborate explanation in the next section.
+
+The `parameters` subcomponent is a list of identifiers that are to be looked-up
+in the symbol-table for their values.
```

Comment:
Is "values" the most appropriate term here?

Reply:
"value" is a word which is a little too general / overloaded in this context. I have gone for "definition" instead for the moment.